File: 2020/day03/toboggan_trajectory.py
import operator
from functools import reduce
from pathlib import Path

import pytest


def trees_encountered(right: int, down: int, tree_map: str) -> int:
    rows = tree_map.strip().split("\n")
    width = len(rows[0])
    x = 0
    trees = 0
    for y, row in enumerate(rows):
        if y % down != 0:
            continue
        position = row[x % width]
        if position == "#":
            trees += 1
        x += right
    return trees


TEST_TREE_MAP = """..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#
"""


@pytest.mark.parametrize("right,down,expected", [
    (1, 1, 2),
    (3, 1, 7),
    (5, 1, 3),
    (7, 1, 4),
    (1, 2, 2),
])
def test_tree_map(right, down, expected):
    assert trees_encountered(right, down, TEST_TREE_MAP) == expected


def main():
    input_file = Path(__file__).parent / "input"
    tree_map = input_file.read_text()

    # part 1
    print(trees_encountered(3, 1, tree_map))

    # part 2
    product = reduce(operator.mul, [
        trees_encountered(1, 1, tree_map),
        trees_encountered(3, 1, tree_map),
        trees_encountered(5, 1, tree_map),
        trees_encountered(7, 1, tree_map),
        trees_encountered(1, 2, tree_map),
    ], 1)
    print(product)


if __name__ == "__main__":
    main()

File: 2020/day03/toboggan_trajectory.nim
import strutils, sequtils, sugar


type
    Slope = object
        x*: Positive
        y*: Positive


const tree = '#'


proc trees_encountered(slope: Slope, tree_map: string): int =
    let rows = tree_map.strip().split({'\n'})
    let width = rows[0].len
    var xpos = 0
    var ypos = 0
    while ypos < rows.len:
        let row = rows[ypos]
        let obstacle = row[xpos mod width]
        if obstacle == tree:
            result += 1
        xpos += slope.x
        ypos += slope.y
    result


# -- part 1


let tree_map = readFile("input")

echo trees_encountered(Slope(x: 3, y: 1), tree_map)


# -- part 2


let result = @[Slope(x: 1, y: 1),
               Slope(x: 3, y: 1),
               Slope(x: 5, y: 1),
               Slope(x: 7, y: 1),
               Slope(x: 1, y: 2)]
    .map(slope => trees_encountered(slope, tree_map))
    .foldl(a * b, 1)

echo result
